app/assets/javascripts/discourse/widgets/quick-access-bookmarks.js.es6

```diff
@@ -0,0 +1,59 @@
+import { h } from "virtual-dom";
+import DiscourseURL from "discourse/lib/url";
+import QuickAccessPanel from "discourse/widgets/quick-access-panel";
+import UserAction from "discourse/models/user-action";
+import { ajax } from "discourse/lib/ajax";
+import { createWidgetFrom } from "discourse/widgets/widget";
+import { postUrl } from "discourse/lib/utilities";
+
+const ICON = "bookmark";
+
+let staleItems = [];
```

Comment:
My feeling is this data would be better stored in the state of a widget, since its value is populated based on the result of the ajax call. Putting the variable here in the closure makes it a bit leaky in tests.  Same for `emptyPlaceholderItemText` below.

Reply:
Totally! I have moved `emptyStatePlaceholderItemText` to the widget state instead.

However, for stale items, I am under the impression that the widget instance is destroyed when "unmounted"? So,

1. Open quick access panel for notifications, and `staleItems` is saved to state.
2. Switch to quick access panel for bookmarks. The notifications widget along with its state (i.e. `staleItems`) is destroyed.
3. Switch back to quick access for notifications. A new notifications widget is initialized --`staleItems` is empty and the loading circle is shown while fresh items are being fetched.

Reply:
After having thought about this a bit more, if we can store `items` as some kind of static variable on the widget, we can drop `staleItems` (and `findStaleItems`) altogether. Because `items === staleItems` when switching back to a quick access panel.

What do you think of:

1. A single object in the base abstract widget's closure. 

    ```js
    const staleItems = {};

    createWidget("quick-access-panel", {
      html() {
        const items = staleItems[this.key] || [];
      }
    });
    ```

2. Emulate static variable with a property on a widget method, like:

    ```js
    createWidget("quick-acces-panel", {
      staleItems() {},

      html() {
        const items = this.staleItems[this.key] || [];
      }
    });
    ```

Reply:
Oh I see, `staleItems` is meant to remember previously seen things even when the widget is destroyed. In that case you should store your data in `Session`. It is a singleton meant for transient data and is cleared between every test run.

Reply:
Ah nice, TIL about the `Session` singleton. Thanks!